iroha_p2p/src/peer.rs

```diff
@@ -139,37 +150,53 @@ where
     /// # Errors
     /// If `new_inner()` errors.
     pub fn new_to(id: PeerId, broker: Broker) -> Result<Self, Error> {
-        Self::new_inner(id, 0_u64, None, State::Connecting, broker)
+        let connection_id = rand::random();
+        Self::new_inner(id, connection_id, None, State::Connecting, broker)
     }
 
-    fn stream(mut read: OwnedReadHalf) -> impl Stream<Item = MessageResult> + Send + 'static {
+    fn stream(
+        mut read: OwnedReadHalf,
+        mut finish: Receiver<()>,
+    ) -> impl Stream<Item = MessageResult> + Send + 'static {
         stream! {
             loop {
-                if let Err(e) = read.as_ref().readable().await {
-                    yield MessageResult::new_error(Error::Io(e));
-                    break;
+                tokio::select! {
+                    readable = read.as_ref().readable() => {
+                        if let Err(e) = readable {
+                            yield MessageResult::new_error(Error::Io(e));
+                            break;
+                        }
+                        let result = match read_message(&mut read).await {
+                            Ok(message) => MessageResult::new_message(message),
+                            Err(e) => {
+                                yield MessageResult::new_error(e);
+                                break;
+                            }
+                        };
+                        yield result;
+                    }
+                    _ = (&mut finish) => {
+                        info!("Connection stream finished");
+                        break;
+                    }
+                    else => break,
                 }
-                let result = match read_message(&mut read).await {
-                    Ok(message) => MessageResult::new_message(message),
-                    Err(e) => MessageResult::new_error(e)
-                };
-
-                yield result;
             }
         }
     }
 
     async fn handshake(&mut self) -> Result<(), Error> {
         let state = self.state;
-        debug!(%state, "Attempting handshake");
+        debug!(%state, id = %self.connection_id, addr = %self.id.address, "Attempting handshake");
         match &self.state {
             State::Connecting => self.connect().await?,
             State::ConnectedTo => self.send_client_hello().await?,
             State::ConnectedFrom => self.read_client_hello().await?,
             State::SendKey => self.send_our_public_key().await?,
             State::GetKey => self.read_theirs_public_key().await?,
             State::Ready => warn!("Not doing handshake, already ready."),
-            State::Error => warn!("Not doing handshake in error state."),
+            State::Disconnected => warn!("Not doing handshake, we are disconnected."),
```

Comment:
Do we need to have specific state like disconnect? Peer just shouldnt exist in such way

Reply:
It may exist for one iteration of connection loop.

Reply:
But does it need to exist though? State is supposed to be of smth which is alive, but it is no longer alive, so no need to keep this state, no?

Reply:
And it has another function: when we disconnect by sending `StopSelf` we set the state to `State::Disconnected`, and when reading handler gets the error id doesn't print it.